plugins/entando-plugin-jacms/src/main/java/com/agiletec/plugins/jacms/apsadmin/resource/MultipleResourceAction.java

```diff
@@ -0,0 +1,340 @@
+            for (String fileDescription : getFileDescriptions()) {
+                if (fileDescription.length() > 0) {
+                    List<BaseResourceDataBean> baseResourceDataBeanList;
+                    BaseResourceDataBean resourceFile = null;
+
+                    File file = getFile(index);
+                    if (null != file) {
+                        resourceFile = new BaseResourceDataBean(file);
+                        resourceFile.setFileName(getFileUploadFileName().get(index));
+                        resourceFile.setMimeType(getFileUploadContentType().get(index));
+
+                    } else {
+                        resourceFile = new BaseResourceDataBean();
+                    }
+                    resourceFile.setDescr(fileDescription);
+                    resourceFile.setMainGroup(getMainGroup());
+                    resourceFile.setResourceType(this.getResourceType());
+                    resourceFile.setCategories(getCategories());
+                    baseResourceDataBeanList = new ArrayList<BaseResourceDataBean>();
+                    baseResourceDataBeanList.add(resourceFile);
+                    try {
+                        if (ApsAdminSystemConstants.ADD == this.getStrutsAction()) {
+                            this.getResourceManager().addResources(baseResourceDataBeanList);
+                            this.addActionMessage(this.getText("message.resource.filename.uploaded",
+                                    new String[]{fileUploadFileName.get(index)}));
+                            savedId.add(index);
+                        } else if (ApsAdminSystemConstants.EDIT == this.getStrutsAction()) {
+                            resourceFile.setResourceId(super.getResourceId());
+                            this.getResourceManager().updateResource(resourceFile);
+                        }
+
+                    } catch (ApsSystemException ex) {
+                        hasError = true;
+                        logger.error("error loading file {} ", fileUploadFileName.get(index), ex);
+                        this.addFieldError(String.valueOf(index), this.getText("error.resource.filename.uploadError",
+                                new String[]{fileUploadFileName.get(index)}));
+                    }
+
+                }
+
+                index++;
+
+            }
+        } catch (Throwable t) {
+            logger.error("error in save", t);
+            return FAILURE;
+        }
+        if (hasError) {
+            logger.error("error uploading one or more resources");
+            savedId.forEach(id -> fileDescriptions.remove(id.toString()));
+            return FAILURE;
+        }
+        return SUCCESS;
+    }
+
+    public String getFileDescription(int i) {
+        if (null != fileDescriptions
+                && !fileDescriptions.isEmpty()) {
+            return fileDescriptions.get(i);
+        }
+        return "";
+    }
+
+    protected void fetchFileDescriptions() {
```

Comment:
If possible fileDesciprtions should be a Map so that all of the loops in this code can be turned into lookups based on file id or file name. 

Also this method should return the value rather than generating global state on the object. Then just call it once and pass the collection around if needed. Setting the global state in a void method leads to problems and lack of clarity.